main/src/library/BagMultiMap.flix

```diff
@@ -0,0 +1,595 @@
+        updateWithKey((_, v1) -> f(v1), k, m)
+
+    ///
+    /// Updates `m` with `k => v1` if `k => v` is in `m` and `f(k, v) = Some(v1)`. Otherwise, returns `m`.
+    ///
+    pub def updateWithKey(f: (k, v) -> Option[v] & ef, k: k, m: BagMultiMap[k, v]): BagMultiMap[k, v] & ef with Order[k] =
+        let BagMultiMap(m1) = m;
+        let outer = (k1, vs) -> if (k1 == k) Nec.map(v1 -> Option.getWithDefault(v1, f(k1, v1)), vs) else vs;
+        BagMultiMap(Map.mapWithKey(outer, m1))
+
+    ///
+    /// Removes the mapping `k` from the BagMultiMap `m`.
+    ///
+    pub def remove(k: k, m: BagMultiMap[k, v]): BagMultiMap[k, v] with Order[k] =
+        let BagMultiMap(m1) = m;
+        BagMultiMap(Map.remove(k, m1))
+
+    ///
+    /// Alias for `findLeft`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def find(f: (k, v) -> Bool, m: BagMultiMap[k, v]): Option[(k, v)] = findLeft(f, m)
+
+    ///
+    /// Optionally returns the first mapping of `m` that satisfies the predicate `f` when searching from left to right.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def findLeft(f: (k, v) -> Bool, m: BagMultiMap[k, v]): Option[(k, v)] =
+        let BagMultiMap(m1) = m;
+        let findGroup = (k, vs) -> Nec.exists(f(k), vs);
+        let findItem = (k, vs) -> Nec.findLeft(f(k), vs) |> Option.map(v1 -> (k, v1));
+        match Map.findLeft(findGroup, m1) {
+            case None          => None
+            case Some((k, vs)) => findItem(k, vs)
+        }
+
+    ///
+    /// Optionally returns the first mapping of `m` that satisfies the predicate `f` when searching from right to left.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def findRight(f: (k, v) -> Bool, m: BagMultiMap[k, v]): Option[(k, v)] =
+        let BagMultiMap(m1) = m;
+        let findGroup = (k, vs: Nec[v]) -> Nec.exists(f(k), vs);
+        let findItem = (k, vs: Nec[v]) -> Nec.findRight(f(k), vs) |> Option.map(v1 -> (k, v1));
+        match Map.findRight(findGroup, m1) {
+            case None          => None
+            case Some((k, vs)) => findItem(k, vs)
+        }
+
+    ///
+    /// Returns a map of all mappings `k => v` in `m` where `v` satisfies the predicate `f`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def filter(f: v -> Bool, m: BagMultiMap[k, v]): BagMultiMap[k, v] with Order[k] = filterWithKey((_, v) -> f(v), m)
+
+    ///
+    /// Returns a map of all mappings `k => v` in `m` where `(k, v)` satisfies the predicate `f`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def filterWithKey(f: (k, v) -> Bool, m: BagMultiMap[k, v]): BagMultiMap[k, v] with Order[k] =
+        foldLeftWithKey((acc, k, v) -> if (f(k, v)) insert(k, v, acc) else acc, empty(), m)
+
+    ///
+    /// Returns a map with mappings `k => f(v)` for every `k => v` in `m`.
+    ///
+    /// Purity reflective: Runs in parallel when given a pure function `f`.
```

Comment:
We have an annotation to express this: `@ParallelWhenPure`

Reply:
Added anno